### ISSUE TITLE
As a data entry clerk I want to access, create, edit people in a place that I can access, and report about them, but not see any of the reports created by others for those contacts.

### ISSUE THREAD

**Issue by senseibara** [medic/cht-core#6352]
**Is your feature request related to a problem? Please describe.**
Currently it doesn't seem possible to limit the replication depth of a user in a way we only have access to contacts of the level below and not to the reports that are submitted on that level below. 

**Describe the solution you'd like**
user 1 : `health_center` level ( primary contact )
user 2 : `clinic` level (primary contact )

Requirement 1 : User 1 should be able to access create and edit people in user's 2 place and be able to submit a report about them. Those reports and contacts should get replicated and accessible to user 2 

Requirement 2 : Reports that User 2 will submit on contacts he have access to should not be replicated to user 1 unless the report include a true `needs_signoff`

**Describe alternatives you've considered**
Right now we are applying a replication depth of 3 for User 1 for him to have requirement 1 but he is seeing all reports that user 2 is submitting which is not what we want. 

**Additional context**
<img width="1147" alt="Capture d’écran 2020-04-10 à 13 04 03" src="https://user-images.githubusercontent.com/24627012/78993251-d32de800-7b2c-11ea-8db2-b3b99b39660a.png">
<img width="1176" alt="Capture d’écran 2020-04-10 à 13 04 42" src="https://user-images.githubusercontent.com/24627012/78993258-d7f29c00-7b2c-11ea-8a2d-d895d4d5e8f0.png">
<img width="392" alt="Capture d’écran 2020-04-10 à 13 05 09" src="https://user-images.githubusercontent.com/24627012/78993259-d9bc5f80-7b2c-11ea-9d68-d21e1813cb4e.png">

**Comment by garethbowen** [comment]
@senseibara Can you provide more information about _why_ user 1 shouldn't see these reports? Do they contain sensitive health information, is it a performance issue, or is user 1 simply not interested in the reports?

**Comment by senseibara** [comment]
@garethbowen for now this is just for us to anticipate performance issues. User 2 might generate up to 18 reports per workflow completed per contact, also we do not expect user 1 to be interested in the reports.

**Comment by MaxDiz** [comment]
Relevant slack conversation: https://medic.slack.com/archives/C3TJFFKK7/p1587582129057200

Adding to v3.10 for MSF-Niger project covid response, and Palladium HIV interventions. Before starting, review prototype feedback with @kennsippell and @senseibara

**Comment by MaxDiz** [comment]
This issue was de-prioritized for its initial intended deployment to MSF-Niger. It may still be relevant for Palladium. I will follow-up in slack with @derickl and @kennsippell to discuss.

**Comment by MaxDiz** [comment]
This issue remains a priority for v3.10. Relevant discussion here: https://medic.slack.com/archives/C3TJFFKK7/p1588706985024100. @derickl please add Palladium feedback as well, so we can make sure we are addressing all related needs in this ticket.

**Comment by kennsippell** [comment]
> User 1 should be able to access create and edit people in user's 2 place and be able to submit a report about them. Those reports and contacts should get replicated and accessible to user 2

Bara, can you clarify this comment? 

> be able to submit a report about them

Who exactly is "them" in that comment. Are you saying User 1 should be able to "submit a report about [a person in user 2's place]"?

**Comment by senseibara** [comment]
That's correct. Right now with Replication depth 1, User 1  will see the place of user 2 but not people/persons created on that place. Ideally we should be able to see these people/persons and submit a report about them that will get replicated to user 2 account, but without seeing reports submitted by user 2 for these people/persons at his level.

**Comment by kennsippell** [comment]
Awesome! As @senseibara describes it, this issue is very useful for addressing the scaling needs of our MSF-Niger covid response and should remain in 3.10.

It would be nice to have somebody confirm the feasibility of this feature request from the product perspective.

**Comment by dianabarsan** [comment]
@kennsippell 
> scaling needs of our MSF-Niger

Are we discussing device-level scaling or server-level scaling? Is it so User1 doesn't generate tasks for User2's reports? ( we don't pass the user object to tasks, right? we have no way of filtering tasks by the user's roles. )

Because of how we do doc authorization, off the top of my head, we'd still need to get the list of all reports submitted under every level the user has access to, just to filter "some" of them to the ones that the user himself has created.

**Comment by kennsippell** [comment]
@dianabarsan 

> we don't pass the user object to tasks, right? we have no way of filtering tasks by the user's roles

Unsure if this is what you're asking, but we [pass the user object](https://github.com/medic/cht-core/blob/master/shared-libs/rules-engine/src/rules-emitter.js#L48) as a global into nools. So tasks can have `appliesIf: () => user.roles.blah`.

This is for device-level scaling. @senseibara Can you work with Diana to clarify needs?

**Comment by dianabarsan** [comment]
I went with adding a new option to `replication_depth` called `report_depth` (final name is TBD) which works in conjuction with the already existent `depth` (see https://github.com/medic/medic-docs/blob/54c9c40c4a2a3c7cc2d0449e05a14ccf3574a5a1/configuration/replication.md#depth)
Ex:
```
{
  "replication_depth": [
    { "role": "district_manager", "depth": 1 },
    { "role": "national_manager", "depth": 2, "report_depth": 1 }
  ]
}
```

In terms of what is visible, for example for 
`{ "role": "national_manager", "depth": 2, "report_depth": 1 }`, a hierarchy like `district (0) > center (1) > clinic (2) > area (3)` and a supervisor at `district` level. 

This supervisor would see `district`, `center` and `clinic`, along with any other person that belongs to `district` and `center`. 
The supervisor would not see any person belonging to `clinic` and no `area`s. 
The `report_depth` of 1, in this case, would mean that this supervisor can only see reports of `district` (0) and `center` (1) level, and no reports at `clinic` level, unless they're submitted by himself! 

If you set `report_depth` to be larger than `depth`, it would have no effect. 
Also, without the `depth` setting, it would have no effect (`{ "role": "district_manager", "report_depth": 1 }` is ignored)

Reports with `needs_signoff` work like before. 

Does that work for you, @senseibara  ?

**Comment by senseibara** [comment]
hey Diana, thanks you first of all this looks really great. 
Will it be possible also to have a 'report_depth' : 0 ? That would mean the supervisor can only see reports of district(0) but won't see reports submitted by center(1) unless those reports are submitted by himself.

**Comment by dianabarsan** [comment]
>Will it be possible also to have a 'report_depth' : 0 ?

Yes, it's possible.

**Comment by senseibara** [comment]
Perfect for me then, nothing else to add on my side. Thank you !

**Comment by garethbowen** [comment]
Ready for AT in `6352-adds-report-replication-depth`, [documentation](https://github.com/medic/cht-docs/pull/58).

**Comment by ngaruko** [comment]
LGTM .
Did you want to have a second look @senseibara

**Comment by dianabarsan** [comment]
@senseibara are you interested to AT this before it gets merged?

**Comment by senseibara** [comment]
sorry @ngaruko @dianabarsan for my late answer, 
I can try this out monday and revert to your ( hoping that's a fine turnaround )

**Comment by senseibara** [comment]
@ngaruko @dianabarsan , i tested a few scenarios and it looks really great to me !  nice work @dianabarsan.

**Comment by dianabarsan** [comment]
Merged into `master`.
